### PR TITLE
Implement DFW settings Exclusion list

### PR DESCRIPTION
**Pull request by ksamoray**
Fixes: #757

### CODE REVIEW

File: nsxt/resource_nsxt_policy_firewall_exclude_list_member.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+func resourceNsxtPolicyFirewallExcludeListMember() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceNsxtPolicyFirewallExcludeListMemberCreate,
+		Read:   resourceNsxtPolicyFirewallExcludeListMemberRead,
+		Delete: resourceNsxtPolicyFirewallExcludeListMemberDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: map[string]*schema.Schema{
+			"member": {
```
Comment (annakhm) [inline nsxt/resource_nsxt_policy_firewall_exclude_list_member.go:25]: Would it make more sense to implement the list as one resource (with list of paths inside?)

Comment (ksamoray) [reply nsxt/resource_nsxt_policy_firewall_exclude_list_member.go:25]: We've considered - But then we'll run into the usual issue where this singleton NSX object is virtually indestructible using NSX API - TF doesn't really like these, I think. Is there any decent way to work around this?

Comment (ksamoray) [reply nsxt/resource_nsxt_policy_firewall_exclude_list_member.go:25]: We can - the question is if there is any advantage in handling this as a scalar, vs handling as a list which is more "natural" to terraform with full CRUD operation.
We've done this for context profile custom attributes and such.

**Comment by ksamoray** [comment]
test-all

File: nsxt/resource_nsxt_policy_firewall_exclude_list_member.go (hunk 1 above)
Comment (annakhm) [reply nsxt/resource_nsxt_policy_firewall_exclude_list_member.go:25]: My personal feeling is that a single string does not justify a dedicated resource. But I don't insist.
As long as we're sure order is not important in this list, we can do it as suggested here

**Review by annakhm — APPROVED** [review]

File: nsxt/resource_nsxt_policy_firewall_exclude_list_member.go (hunk 1 above)
Comment (ksamoray) [reply nsxt/resource_nsxt_policy_firewall_exclude_list_member.go:25]: @annakhm there is a more serious problem which I should look into:
When two members are created at the same time, there is a concurrency problem - both retrieve the member list, append the new member and when submitting one change is overwritten.
I'm not sure if we have some mean to handle this (and we might have this elsewhere maybe).

Comment (ksamoray) [reply nsxt/resource_nsxt_policy_firewall_exclude_list_member.go:25]: I worked around the concurrency problem, and added a few test. I'd rather if you could have another look...

File: nsxt/resource_nsxt_policy_firewall_exclude_list_member.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+func resourceNsxtPolicyFirewallExcludeListMemberDelete(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	member := d.Get("member").(string)
+
+	doUpdate := func() error {
+		var obj model.PolicyExcludeList
+
+		client := security.NewExcludeListClient(getSessionContext(d, m), connector)
+		obj, err := client.Get()
+		if isNotFoundError(err) {
```
Comment (annakhm) [inline nsxt/resource_nsxt_policy_firewall_exclude_list_member.go:124]: We should silently swallow the error in `not found` case

**Comment by annakhm** [comment]
I still feel like a singleton resource is better design, as it would save roundtrips and avoid concurrency problems. But I don't insist.

**Comment by ksamoray** [comment]
> I still feel like a singleton resource is better design, as it would save roundtrips and avoid concurrency problems. But I don't insist.

IMO a singleton doesn't make sense in the API level as well, from the functional aspect, e.g:
- Why should this resource be managed from a single point?
- Should a user be able to create dependency between a exclude list member and other resources?
- Why would we confine an organization to maintain the exclude list from a single interface (UI, TF...)? As otherwise concurrency issues become worse.

**Comment by salv-orlando** [comment]
@ksamoray shouldn't this PR also have doc changes since we are adding a new resource?

Regarding the singleton vs non-singleton use case, I think both @annakhm  and @ksamoray have good points.
I'd tend to slightly prefer user experience over scalability, especially considering we should not see hundreds or thousands of groups in the exclusion list, so I'd say maybe concurrency is not a big concern.

**Comment by ksamoray** [comment]
@salv-orlando yeah we need documentation but no point in adding that before we concluded the implementation.

**Review by annakhm — APPROVED** [review]

**Pull request merged**